segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/LogReader.java

```diff
@@ -185,14 +188,72 @@ public String toString() {
 
     //region ReadLedger
 
-    @RequiredArgsConstructor
+    private final static class EntryHolder {
+        final long entryId;
+        final int length;
+        final InputStream entryContent;
+
+        private EntryHolder(LedgerEntry entry) {
+            this.entryId = entry.getEntryId();
+            this.length = entry.getEntryBuffer().readableBytes();
+            // This call doesn't change the reference count on the returned bytebuf
+            final ByteBuf internalBuffer = entry.getEntryBuffer();
+            // the refcount will be decremented in LogReader#release
+            this.entryContent = new ByteBufInputStream(internalBuffer, false /* releaseOnClose */);
+        }
+    }
+
     private static class ReadLedger {
         final LedgerMetadata metadata;
-        final LedgerHandle handle;
-        final Enumeration<LedgerEntry> reader;
+        final ReadHandle handle;
+        final LedgerEntries ledgerEntries;
+        /**
+         * Holder for the list of entries.
+         * A null value means that this ledger is empty.
+         */
+        final Iterator<EntryHolder> entryIterator;
+        final AtomicBoolean closed = new AtomicBoolean(false);
+
+        public ReadLedger(LedgerMetadata metadata, ReadHandle handle, LedgerEntries ledgerEntries) {
+            this.metadata = metadata;
+            this.handle = handle;
+            this.ledgerEntries = ledgerEntries;
+            if (ledgerEntries != null) {
+                List<EntryHolder> entries = new ArrayList<>();
```

Comment:
I still don't understand why we need to loop through all the entries here, convert them to `EntryHolder` instances, then in `nextItem` we just loop through this array. We are still holding a pointer to `ledgerEntries`.

Can't we just have `nextItem` use the `ledgerEntries.iterator()` and return from there? That way we won't have a need for `EntryHolder` nor will we need to double-loop through the entries.

Reply:
Good idea!
Current code is the result of the first implementation with a the memcopy, now this nee ArrayList is a waste.
I will update the patch

Reply:
Good idea!
Current code is the result of the first implementation with a the memcopy, now this nee ArrayList is a waste.
I will update the patch